src/derivative_wrappers.jl

```diff
@@ -20,45 +20,58 @@ function derivative!(df::AbstractArray{<:Number}, f, x::Union{Number,AbstractArr
     nothing
 end
 
+jacobian_autodiff(f,x,_,_)=ForwardDiff.derivative(f,x)
+jacobian_autodiff(f,x::AbstractArray,_,hascolor::Type{Val{false}})=ForwardDiff.jacobian(f,x)
+function jacobian_autodiff(f,x::AbstractArray,integrator,hascolor::Type{Val{true}})
+  colorvec=integrator.f.colorvec
+  sparsity=integrator.f.jac_prototype
+  jac=integrator.f.jac
+  J=jac isa SparseMatrixCSC ? similar(jac) : zeros(size(jac)...)
+  forwarddiff_color_jacobian!(J,f,x,color=colorvec,sparsity=sparsity)
+  J
+end
+jacobian_finitediff(f,x,difftype,_,_)=DiffEqDiffTools.finite_difference_derivative(f, x, difftype, eltype(x))
+jacobian_finitediff(f,x::AbstractArray,difftype,_,hascolor::Type{Val{false}})=DiffEqDiffTools.finite_difference_jacobian(f, x, difftype, eltype(x), Val{false})
+jacobian_finitediff(f,x::AbstractArray,difftype,integrator,hascolor::Type{Val{true}})=
+  DiffEqDiffTools.finite_difference_jacobian(f, x, difftype, eltype(x), Val{false},color=integrator.f.colorvec,sparsity=integrator.f.jac_prototype)
+
 function jacobian(f, x,
                   integrator::DiffEqBase.DEIntegrator)
     local J
     alg = unwrap_alg(integrator, true)
-    isarray = typeof(x) <: AbstractArray
     if get_current_alg_autodiff(integrator.alg, integrator.cache)
-      if isarray
-        J = ForwardDiff.jacobian(f,x)
-      else
-        J = ForwardDiff.derivative(f,x)
-      end
+      J = jacobian_autodiff(f,x,integrator,Val{DiffEqBase.has_colorvec(integrator.f)})
```

Comment:
I'm wondering if introducing those functions and dispatching is useful at all. Although I think that should not put too much stress on the compiler I'm looking for every tiny bit that could possibly explain a slowdown :stuck_out_tongue: I just read through the documentation (https://docs.julialang.org/en/v1/manual/types/index.html#%22Value-types%22-1 and https://docs.julialang.org/en/v1/manual/performance-tips/#The-dangers-of-abusing-multiple-dispatch-(aka,-more-on-types-with-values-as-parameters)-1) and I'm worried that it's not useful at all to dispatch here. Additionally, according to the documentation one should rather use `Val(...)` instances for consistency :open_mouth: I had never heard this before.

Reply:
Well that's suggested by @ChrisRackauckas to reduce runtime brunch.  And yeah,  I should use Val(),  sorry I didn't read docs carefully.

Reply:
Yes, this is an inproper use of `Val` though. The chunksize needs to be saved as a type, otherwise this will dynamic dispatch every time and be slow unless it infers. Why does this part of the code make a `Val`?

I don't think this is blocking though and can get cleaned up later.